### DISCUSSION TITLE
C++ question, templates...

### DISCUSSION THREAD

**Discussion by csccva** [AdaptiveCpp/AdaptiveCpp#1315] [General]
Hello, 

This is more a C++ question. 

I took this code from the book Data Parallel C++ (2nd ed.)
```
// Copyright (C) 2023 Intel Corporation

// SPDX-License-Identifier: MIT

#include <cstdio>
#include <numeric>
#include <sycl/sycl.hpp>

using namespace sycl;

int main() {
  constexpr size_t N = 16;
  constexpr size_t B = 4;

  queue q;
  int* data = malloc_shared<int>(N, q);
  int* sum = malloc_shared<int>(1, q);
  std::iota(data, data + N, 1);
  *sum = 0;

  // BEGIN CODE SNIP
  q.parallel_for(nd_range<1>{N, B}, [=](nd_item<1> it) {
     int i = it.get_global_id(0);
     auto grp = it.get_group();
     int group_sum =
         reduce_over_group(grp, data[i], plus<>());
     if (grp.leader()) {
       atomic_ref<int, memory_order::relaxed,
                  memory_scope::system,
                  access::address_space::global_space>(
           *sum) += group_sum;
     }
   }).wait();
  // END CODE SNIP

  std::cout << "sum = " << *sum << "\n";
  bool passed = (*sum == ((N * (N + 1)) / 2));
  std::cout << ((passed) ? "SUCCESS" : "FAILURE") << "\n";

  free(sum, q);
  free(data, q);
  return (passed) ? 0 : 1;
}
``` 

I compiled with OneApi and it seemed to have worked (compilation and running on CPU and Nvidia). 
When I compile it with AdaptiveCpp  (v0.9.4 installed via spack on nvidia machine) I get this error:
```
reduction.cpp:26:42: error: too few template arguments for class template 'plus'
         reduce_over_group(grp, data[i], plus<>());
                                         ^
/scratch/project_2008874/spack/opt/spack/linux-rhel8-x86_64_v3/gcc-11.2.0/hipsycl-0.9.4-qmrshzibdb5uzqdhetqwioszhe3pgurw/bin/../include/sycl/../hipSYCL/sycl/libkernel/functional.hpp:37:30: note: template is declared here
template <typename T> struct plus {
```
If I change `plus<>()` to `plus<int>()` both OneApi and AdaptiveCpp compile and work.

I understand the basic idea of templates, but I am not good at C++ and  I would like to understand what happens.

Cristian

**Comment by illuhad** [comment]
This is because the functional objects like `plus<T>` were given a default argument of `plus<class T = void>` in the final version of the spec. We have a PR somewhere that implements, IIRC, but it's not merged.

**Comment by csccva** [comment]
Thanks for the reply. I  understand. 
I am preparing some codes/exercises for a training and trying to make at least the basic examples as portable as possible between AdapticCpp and OneAPI.

Cristian
